tools/crm_mon.c

```diff
@@ -1310,8 +1338,20 @@ reconcile_output_format(pcmk__common_args_t *args) {
         /* Default case.
          * No explicit format given and none of the above overrides were hit.
          */
-        pcmk__str_update(&args->output_ty, "console");
-        output_format = mon_output_console;
+#if CURSES_ENABLED
+        if (args->version) {
+            // We don't need to initialize curses to display the version
+            options.one_shot = TRUE;
+            retval = no_curses_cb("-N", NULL, NULL, &error);
```

Comment:
> In any case we'd only want to do this if console mode -- version works fine with none/html/xml.

What do you mean by this? XML enables one-shot mode if we're not in daemonized mode.  It turns out that HTML doesn't. It just prints once to the screen and then never updates or returns to a prompt:
```
[reid@laptop pacemaker]$ sudo crm_mon --output-as=html
<!DOCTYPE html>
...
<body>
<span>Not connected to CIB</span><span>: </span><span>Waiting for initial connection</span>
</body>
</html>


```

Of course "none" is ignored completely for now, as we identified in other comments. My preference is to have HTML and none formats also default to one-shot.

Reply:
And interestingly, when I remove this condition (specifically its equivalent on the main branch) so that `crm_mon` uses `curses_ver()`, it's fine and quick but it gives this extra indentation that's not in the `curses_ver()` code:
```
[reid@laptop pacemaker]$ sudo crm_mon --version
Pacemaker 2.1.5
               Written by Andrew Beekhof and the Pacemaker project contributors
```

Reply:
> > In any case we'd only want to do this if console mode -- version works fine with none/html/xml.
> 
> What do you mean by this?

We don't want to call no_curses_cb() if xml or html were given, but I missed the context where they've already been filtered out

Before the output format stuff, --version ignored all other arguments and just printed to stdout. With the new style, the version should be printed in the requested format. (And of course return immediately.)

> XML enables one-shot mode if we're not in daemonized mode. It turns out that HTML doesn't. It just prints once to the screen and then never updates or returns to a prompt:
> 
> ```
> [reid@laptop pacemaker]$ sudo crm_mon --output-as=html
> <!DOCTYPE html>
> ...
> <body>
> <span>Not connected to CIB</span><span>: </span><span>Waiting for initial connection</span>
> </body>
> </html>
> ```
> 
> Of course "none" is ignored completely for now, as we identified in other comments. My preference is to have HTML and none formats also default to one-shot.

Originally, HTML did not default to one-shot. Not sure whether that ever worked, though. Any idea why it doesn't update?

--external-agent doesn't require --daemonize, so it would theoretically be reasonable to run crm_mon --output-as=none with an external agent and have it update.

HTML, XML, and none should at least work with update when --interval is explicitly given.

Reply:
> With the new style, the version should be printed in the requested format. (And of course return immediately.)

That does happen, except that for console format where we currently default to text. I'd prefer to continue defaulting to text when `--version` is given with console mode., since we get the weird tab shown above. We can't remove `curses_ver()` but we could make it empty since it's unused.

> Originally, HTML did not default to one-shot. Not sure whether that ever worked, though. Any idea why it doesn't update?

I'll look into it

Reply:
> Originally, HTML did not default to one-shot. Not sure whether that ever worked, though. Any idea why it doesn't update?

Figured that out -- the output object has to finish in order to print anything. Needed this:
```
@@ -2069,7 +2069,7 @@ mon_refresh_display(gpointer user_data)
                                 false, &pcmkd_state);
     }
 
-    if (out->dest != stdout) {
+    if (output_format != mon_output_console) {
         out->reset(out);
     }
 
@@ -2087,7 +2087,7 @@ mon_refresh_display(gpointer user_data)
         return G_SOURCE_REMOVE;
     }
 
-    if (out->dest != stdout) {
+    if (output_format != mon_output_console) {
         out->finish(out, CRM_EX_OK, true, NULL);
     }
```

Now they update, but they don't redraw the screen. Probably because they use `fprintf(out->dest, ...)` instead of `printw()`, etc. They're printing to stdout instead of to the window. Even if we call `clear()` and `refresh()`, the old output doesn't get cleared -- it's like tailing a log file. It simply re-prints the whole output on its own as cluster updates come in.

It's semi-interactive. Pressing `'?'` does nothing. Pressing the other interactive commands works after pressing Enter (for example, 'n' -> Enter toggles grouping by node).

I think that's as good as we're going to get without significant changes to the output library. Now `--output-as={html,xml,none}` work with interactive mode. If neither `--one-shot` nor `--daemonize` is given, these formats default to interactive mode.

This does raise the question of whether we should continue treating "text" specially. Currently, if neither `--one-shot` nor `--daemonize` is given, `--output-as=text` automatically enables one-shot mode. It's been that way for at least a while, so not sure that changing it would be a good idea now. However, if we wanted it to be consistent with HTML and XML, then we would have it print directly to stdout with continuous, non-interactive scrolling updates, instead of using curses. That's typically not very useful -- curses is better -- though it could be useful if curses is unavailable.

Reply:
We don't want interactive or stdout with XML/HTML in combination with updates. Updating should only be allowed with a file.

Reply:
>  I'd prefer to continue defaulting to text when --version is given with console mode

Makes sense

Reply:
> We don't want interactive or stdout with XML/HTML in combination with updates. Updating should only be allowed with a file.

Why not? And I did this because you said:

> Originally, HTML did not default to one-shot.
> HTML, XML, and none should at least work with update when --interval is explicitly given.

If we're not defaulting to one-shot or interactive, then it sounds like that leaves defaulting to daemonized mode if `--output-to` is given and throwing an error if not.

I don't see an issue with allowing XML/HTML updates to go to stdout if the user doesn't specify a destination. The output is correct, and it doesn't require as much special case handling.

Re: "we don't want interactive" specifically -- it's still interactive even when redirected to a file. If you enter ''n'->Enter' where the foreground process is running, the file gets updates with group-by-node toggled. Avoiding that would require more special case handling.

---

Replying to myself:
> Now they update, but they don't redraw the screen. 

I had forgotten that `curses_init()` is what creates the window in the first place. So not only are we not writing to stdscr, we're never initializing curses.

I believe it would be possible to have XML and HTML outputs use curses via a global flag or something and then adding the appropriate stuff (based on `crm_mon_curses.c`) to their output functions. I don't know that it's worth the effort, especially right now.

Reply:
> > Originally, HTML did not default to one-shot.
> > HTML, XML, and none should at least work with update when --interval is explicitly given.
> 
> If we're not defaulting to one-shot or interactive, then it sounds like that leaves defaulting to daemonized mode if `--output-to` is given and throwing an error if not.

HTML/XML and curses are exclusive, never combined. Interactive refers to the curses capability of accepting input, whereas updates (repeatedly updating the output at an interval) can be used without interactivity or curses.

I think always defaulting to one-shot for HTML and XML is the least confusing approach at this point, but we should support updates if --interval is explicitly given. Someone displaying HTML or XML to a file likely wants to serve or parse that file with up-to-date information. (If in production, they'll use --daemonize, but while testing or debugging, they may not.)
 
> I don't see an issue with allowing XML/HTML updates to go to stdout if the user doesn't specify a destination. The output is correct, and it doesn't require as much special case handling.

Without curses involved, stdout is never truly overwritten, in the way that a file is. It's just repeatedly spewed, which makes no sense for HTML or XML (it can't be served or parsed, just viewed by a human). So, defaulting to one-shot for XML or HTML to stdout reflects that. We could default to updates for XML or HTML to a file, and that's probably the original intent, but for consistency we could just default to one-shot in that case as well.
 
> Re: "we don't want interactive" specifically -- it's still interactive even when redirected to a file. If you enter ''n'->Enter' where the foreground process is running, the file gets updates with group-by-node toggled. Avoiding that would require more special case handling.

Interactivity should only be enabled with curses, and I believe that's the case.

> I believe it would be possible to have XML and HTML outputs use curses via a global flag or something and then adding the appropriate stuff (based on `crm_mon_curses.c`) to their output functions. I don't know that it's worth the effort, especially right now.

There's no use case for curses displaying XML or HTML. Curses is strictly for interactive, human-friendly output. You can think of curses as update mode for text to stdout.

Reply:
> Interactivity should only be enabled with curses, and I believe that's the case.

It isn't

Reply:
Well, it is **right now** in main, because we don't support updates for XML and HTML currently. If we do (as I was doing in the most recent iteration of this PR), then interactivity is allowed while updating a file.

Sigh

Reply:
We don't support updates at all, or we support updates only with --daemonize? Updating an HTML file definitely worked at some point in the past; that was the intent of HTML output. A user could run a web server on a cluster node, daemonize crm_mon to output HTML with updates to the document root, and view the current cluster status from any local web browser.

Reply:
I was mistaken, it does work with HTML. I must have typo'd the output file or something.